---
title: "Report"  
output: 
  flexdashboard::flex_dashboard:
    vertical_layout: scroll
    source_code: embed
    orientation: rows
    theme: paper
params:
  tree: "data/nogaps_tree_nogaps_treeannotator_ann.trees"
  neo_species: "T_conf"
  busco_fig: "results/busco_figure.png"
---

```{r setup, include=FALSE, warning=FALSE}
require("ggplot2")
require("tidyverse")
require("plotly")
require("ape")
require("flexdashboard")
require("rcompanion")
```

Dosage Balance and Compensation
=======================================================================

Row {data-height=650}{.tabset .tabset-fade}
-----------------------------------------------------------------------


### Dosage Balance Results of X chromosome

<!-- The following graph shows the ratio of X to Autosome expression profile for males and females. -->

```{r balance, echo=FALSE, fig.height=5, fig.width=8}
dataFiles <- lapply(Sys.glob("../results/*_TMM_ann.txt"), read.table)
all_species_df = data.frame()
count <- 1
for (file in dataFiles){
  # file <- dataFiles[[1]]
  colnames(file) <- c("seq_id", "sample", "TMM","gender","species","chrom_num","chrom_name", "og")
  file <- file[file$chrom_name != "Unknown" & file$chrom_name != "Un",]
  # Find Autosome median expression in each species
  autosome_median <- file %>%
    filter(chrom_name == "Autosome") %>%
    summarise(med = median(TMM, na.rm = T)) %>%
    pull(med)
  
  # Find median expression of X,Neo-X and autosome in each species
  file <- file %>%
    group_by(chrom_name) %>%
    mutate(chrom_name_median=median(TMM))
  
  file <- file %>%
  filter(chrom_name == "Neo-X" | chrom_name == "X") %>%
  group_by(chrom_name,gender) %>%
  mutate(X_AA_ratio=log2(TMM/autosome_median))
  
  # Combine each species dataframe to all species dataframe.
  all_species_df <- rbind(as.data.frame(all_species_df),as.data.frame(file))
  count <- count + 1
}

plot_data_x <- groupwiseMedian(X_AA_ratio ~ species + gender,
                             data = all_species_df[all_species_df$chrom_name == "X",],
                             bca = FALSE,
                             conf = 0.95,
                             percentile = TRUE)
plot_data_x$species_gender <- paste(plot_data_x$species,plot_data_x$gender)

plot_data_neox <- groupwiseMedian(X_AA_ratio ~ species + gender,
                               data = all_species_df[all_species_df$chrom_name == "Neo-X",],
                               bca = FALSE,
                               conf = 0.95,
                               percentile = TRUE)
plot_data_neox$species_gender <- paste(plot_data_neox$species,plot_data_neox$gender)

balance <- ggplot(plot_data_x, aes(x=Median, y=species_gender,xmin=Percentile.lower,xmax=Percentile.upper, color=gender)) + 
  # geom_boxplot() +
  geom_point(aes(size=1))+
  scale_colour_manual(name="",  
                      values = c("Male"="#E69F00",
                                 "Female" = "#0072B2"))+
  geom_errorbarh(height = 0.01)+
  geom_vline(aes(xintercept = -1)) +
  geom_vline(aes(xintercept = 1)) +
  guides(fill=FALSE) + 
  coord_cartesian(xlim = c(-2,2))+
  guides(size=F)+
  xlab("log2(X to Autosome) Ratio") +
  ylab("Species")+
  labs(color = "Gender")+
  facet_wrap(~species,nrow = 5, scales="free_y", strip.position = "left")+
  theme_minimal()+
  theme(axis.text.y=element_blank(),
        axis.text.x=element_text(size = 12),
        axis.ticks.y=element_blank(),
        strip.text.y = element_text(size = 14, angle = 180),
        axis.title = element_text(size = 14))
# ggplotly(balance)
balance
```

### Dosage Balance Results of Neo-X chromosome

```{r neoxbalance,echo=FALSE, fig.height=5, fig.width=8}
neobalance <- ggplot(plot_data_neox, aes(x=Median, y=species_gender,xmin=Percentile.lower,xmax=Percentile.upper, color=gender)) + 
  # geom_boxplot() +
  geom_point(aes(size=1))+
  scale_colour_manual(name="",  
                      values = c("Male"="#E69F00",
                                 "Female" = "#0072B2"))+
  geom_errorbarh(height = 0.01)+
  geom_vline(aes(xintercept = -1)) +
  geom_vline(aes(xintercept = 1)) +
  guides(fill=FALSE) + 
  coord_cartesian(xlim = c(-2,2))+
  guides(size=F)+
  xlab("log2(Neo-X to Autosome) Ratio") +
  ylab("Species")+
  labs(color = "Gender")+
  facet_wrap(~species,nrow = 5, scales="free_y", strip.position = "left")+
  theme_minimal()+
  theme(axis.text.y=element_blank(),
        axis.text.x=element_text(size = 12),
        axis.ticks.y=element_blank(),
        strip.text.y = element_text(size = 14, angle = 180),
        axis.title = element_text(size = 14))
# ggplotly(neobalance)
neobalance
```

<!-- The above graph shows the ratio of Neo-X to Autosome expression profile for males and females. -->

### Dosage Compensation Weighted by Ancestral State


```{r weigted, echo=FALSE, fig.height=5, fig.width=8}
# Add chromosome category
all_species_df$chrom_cat <- (ifelse(all_species_df$species == params$neo_species, "Neo-X", "Anc-X"))

# Declare a datafrane for free and weights
species_timetree_df <- data.frame()

# Read the tree
species_timetree <- read.nexus(params$tree)

# Dist between any two species
t <- as.data.frame(cophenetic.phylo(species_timetree))

# Dist between any species to neo-x species
species_timetree_df <- gather(t[params$neo_species,])

# Add headers to the dataframe
colnames(species_timetree_df) <- c("species","divTime_to_neo")

# Calculate relative weights based on divergence time
species_timetree_df <- species_timetree_df %>%
  mutate(weights = 1- (divTime_to_neo/sum(divTime_to_neo)))

# Add weights column to all_species_df
all_species_df$weights <- species_timetree_df$weights [match(all_species_df$species,species_timetree_df$species)]

# Calculate Weighted Expression by multiplying the relative weights with the scaled expression values
all_species_df <- all_species_df %>%
  group_by(species) %>%
  mutate(weighted_expression=TMM*weights)

# Calculate Median Weighted Expression for each orthogroup
all_species_minus_T_conf_df <- all_species_df %>%
  filter(species != params$neo_species) %>%
  group_by(og) %>%
  mutate(med_weighted_expression=median(weighted_expression))

# Calculate Median Weighted Expression for each orthogroup
T_conf_df <- all_species_df %>%
  filter(species == params$neo_species) %>%
  group_by(og) %>%
  mutate(med_weighted_expression=median(weighted_expression))

all_species_df <- rbind(all_species_minus_T_conf_df,T_conf_df)

all_species_df <- all_species_df %>%
  group_by(og,gender) %>%
  mutate(neox_ancx = log2(median(weighted_expression[species == "T_conf"])/median(weighted_expression[species != "T_conf"])))
all_species_df <- all_species_df[is.finite(all_species_df$neox_ancx),]

write.table(all_species_df,file = "../results/all_species_df_R.txt", sep="\t",row.names = F, col.names = T)



# Create a ggplot object
plot_data_ancx <- groupwiseMedian(neox_ancx ~ species + gender,
                                  data = all_species_df[all_species_df$chrom_name == "Neo-X",],
                                  bca = FALSE,
                                  conf = 0.95,
                                  percentile = TRUE)
plot_data_ancx$species_gender <- paste(plot_data_ancx$species,plot_data_ancx$gender)
ancbalance <- ggplot(plot_data_ancx, aes(x=Median, y=species_gender,xmin=Percentile.lower,xmax=Percentile.upper, color=gender)) + 
  # geom_boxplot() +
  geom_point(aes(size=1))+
  scale_colour_manual(name="",  
                      values = c("Male"="#E69F00",
                                 "Female" = "#0072B2"))+
  geom_errorbarh(height = 0.01)+
  geom_vline(aes(xintercept = -1)) +
  geom_vline(aes(xintercept = 1)) +
  guides(fill=FALSE) + 
  coord_cartesian(xlim = c(-2,2))+
  guides(size=F)+
  xlab("log2 (Neo-X to Anc-X) ratio") +
  ylab("Species")+
  labs(color = "Gender")+
  facet_wrap(~species,nrow = 5, scales="free_y", strip.position = "left")+
  theme_minimal()+
  theme(axis.text.y=element_blank(),
        axis.text.x=element_text(size = 12),
        axis.ticks.y=element_blank(),
        strip.text.y = element_text(size = 14, angle = 180),
        axis.title = element_text(size = 14))

# Visualize the plot using plotly
ancbalance
# ggplotly(ancbalance)
```

<!-- The graph below shows the normalized, scaled and log transformed expression of X chromosome, Neo-X and the autosome. Here the weights for Anc-X are reciprocal to the distance from the common ancestor based on the in put time tree. -->


Row
-----------------------------------------------------------------------

### Expression Profile of Transcript in Neo-X vs Anc-X

```{r exp,echo=FALSE}

unique_OG_all_species_df <- all_species_df %>%
  distinct(og,chrom_cat,med_weighted_expression,chrom_num)
ggplot(unique_OG_all_species_df,
       aes(x=chrom_cat,
           y=med_weighted_expression,
           group=factor(og),
           color=chrom_cat))+
  geom_line(alpha=0.2,
            color="black",
            position = "stack")+
  geom_point(alpha=0.2,
             position = "stack")+
  xlab("Chromosome")+
  ylab("Weighted TMM")+
  guides(color=F)+
  theme(legend.position='none')+
  facet_wrap(~gender)+
  theme_minimal()
# ggplotly()
```


<!-- The graph below shows how the (weighted) expression of each gene/transcript/orthogroup in the Neo-X relates to the (weighted) median Anc-X expression  -->

BUSCO  
=======================================================================

Row
-----------------------------------------------------------------------

### BUSCO Results

The following graphs show the before and after of the clustering methods on the trinity transcriptome.

```{r busco}
knitr::include_graphics(params$busco_fig)
```
